app_flutter/lib/status_grid.dart

```diff
@@ -45,29 +44,151 @@ class StatusGridContainer extends StatelessWidget {
           );
         }
 
+        final StatusGridHelper statusGridHelper =
+            StatusGridHelper(statuses: statuses);
         return StatusGrid(
           statuses: statuses,
+          taskMatrix: statusGridHelper.taskMatrix,
+          taskIconRow: statusGridHelper.taskIconRow,
         );
       },
     );
   }
 }
 
+/// Class to handle data operations on [List<CommitStatus>].
+/// 
+/// Flattens the mapping of one [CommitStatus] from many [Stage] objects,
+/// where each [Stage] object maps to many [Task] objects, to a 2D matrix.
+/// 
+// TODO(chillers): Support special ordering of taskMatrix. https://github.com/flutter/cocoon/issues/478
+class StatusGridHelper {
+  StatusGridHelper({@required this.statuses}) {
+    _columnKeyIndex = _createTaskColumnKeyIndex(statuses);
+
+    _taskIconRow = _createTaskIconRow(statuses, _columnKeyIndex);
+    _taskMatrix = _createTaskMatrix(statuses, _columnKeyIndex);
+  }
+
+  final List<CommitStatus> statuses;
+
+  /// A key, value table to find what column a [Task] is in.
+  Map<String, int> _columnKeyIndex;
+
+  /// Computed 2D array of [Task] to make it easy to retrieve and sort tasks.
+  List<List<Task>> get taskMatrix => _taskMatrix;
+  List<List<Task>> _taskMatrix;
+
+  /// A list of [Task] objects that can be used for the row of [TaskIcon].
+  List<Task> get taskIconRow => _taskIconRow;
+  List<Task> _taskIconRow;
+
+  /// A unique index for grouping [Task] from separate [CommitStatus].
+  ///
+  /// [task.stageName] and [task.name] are not unique. However, they
+  /// are unique when combined together.
+  String _taskColumnKey(Task task) {
+    return '${task.stageName}:${task.name}';
+  }
+
+  /// A map of all [taskColumnKey] to a unique index from [List<CommitStatus>].
+  ///
+  /// Scans through all [Task] in [List<CommitStatus>] to find all unique [taskColumnKey].
+  /// This ensures a column is allocated in [_createTaskMatrix].
+  ///
+  /// When a new [taskColumnKey] is found, it is inserted and the index is incremeted.
+  Map<String, int> _createTaskColumnKeyIndex(List<CommitStatus> statuses) {
+    final Map<String, int> taskColumnKeyIndex = <String, int>{};
+    int currentIndex = 0;
+
+    /// O(Tasks * CommitStatuses). In production, this is usually O(85 * 100) ~ 8500 operations.
+    for (CommitStatus status in statuses) {
+      for (Stage stage in status.stages) {
+        for (Task task in stage.tasks) {
+          final String key = _taskColumnKey(task);
+          if (taskColumnKeyIndex.containsKey(key)) {
+            continue;
+          }
+
+          taskColumnKeyIndex[key] = currentIndex++;
+        }
+      }
+    }
+
+    return taskColumnKeyIndex;
+  }
```

Comment:
Just want to mention another lazy style to calculate taskColumnKeyIndex, which returns the index on the fly: 
```
Map<String, int> taskColumnKeyIndex = <String, int>{};
int getColumnKeyIndex(Task task) {
  final String key = _taskColumnKey(task);
  if (taskColumnKeyIndex.containsKey(key)) {
    return taskColumnKeyIndex[key];
  }
  int id = taskColumnKeyIndex.keys.length.
  taskColumnKeyIndex[key] = id;
  return id;
}
```